pkg/redpanda/client.go

```diff
@@ -0,0 +1,210 @@
+	"errors"
+	"fmt"
+	"net"
+	"slices"
+	"strings"
+
+	"github.com/redpanda-data/helm-charts/charts/redpanda"
+	"github.com/redpanda-data/helm-charts/pkg/gotohelm/helmette"
+	"github.com/redpanda-data/helm-charts/pkg/kube"
+	"github.com/twmb/franz-go/pkg/kgo"
+	"github.com/twmb/franz-go/pkg/sasl"
+	"github.com/twmb/franz-go/pkg/sasl/plain"
+	"github.com/twmb/franz-go/pkg/sasl/scram"
+	corev1 "k8s.io/api/core/v1"
+)
+
+var (
+	ErrServerCertificateNotFound          = errors.New("kafka TLS certificate not found")
+	ErrServerCertificatePublicKeyNotFound = errors.New("kafka TLS certificate does not contain a public key")
+
+	ErrClientCertificateNotFound           = errors.New("client TLS certificate not found")
+	ErrClientCertificatePublicKeyNotFound  = errors.New("client TLS certificate does not contain a public key")
+	ErrClientCertificatePrivateKeyNotFound = errors.New("client TLS certificate does not contain a private key")
+
+	ErrSASLSecretNotFound          = errors.New("users secret not found")
+	ErrSASLSecretKeyNotFound       = errors.New("users secret key not found")
+	ErrSASLSecretSuperuserNotFound = errors.New("users secret has no users")
+
+	supportedSASLMechanisms = []string{
+		"PLAIN", "SCRAM-SHA-256", "SCRAM-SHA-512",
+	}
+)
+
+func firstUser(data []byte) (user string, password string, mechanism string) {
+	file := string(data)
+
+	for _, line := range strings.Split(file, "\n") {
+		tokens := strings.Split(line, ":")
+		if len(tokens) != 3 {
+			continue
+		}
+
+		if !slices.Contains(supportedSASLMechanisms, tokens[2]) {
+			continue
+		}
+
+		user, password, mechanism = tokens[0], tokens[1], tokens[2]
+		return
+	}
+
+	return
+}
+
+func saslOpt(user, password, mechanism string) kgo.Opt {
+	var m sasl.Mechanism
+	switch mechanism {
+	case "PLAIN":
+		m = plain.Auth{User: user, Pass: password}.AsMechanism()
+	case "SCRAM-SHA-256", "SCRAM-SHA-512":
+		scram := scram.Auth{User: user, Pass: password}
+
+		switch mechanism {
+		case "SCRAM-SHA-256":
+			m = scram.AsSha256Mechanism()
+		case "SCRAM-SHA-512":
+			m = scram.AsSha512Mechanism()
+		}
+	default:
+		panic(fmt.Sprintf("unhandled SASL mechanism: %s", mechanism))
+	}
+
+	return kgo.SASL(m)
+}
+
+// DialContextFunc is a function that acts as a dialer for the underlying Kafka client.
+type DialContextFunc func(ctx context.Context, network, host string) (net.Conn, error)
+
+func wrapTLSDialer(dialer DialContextFunc, config *tls.Config) DialContextFunc {
+	return func(ctx context.Context, network, host string) (net.Conn, error) {
+		conn, err := dialer(ctx, network, host)
+		if err != nil {
+			return nil, err
+		}
+		return tls.Client(conn, config), nil
+	}
+}
+
+// KafkaClient creates a client to talk to a Redpanda cluster based on its helm
+// configuration over its internal listeners.
+func KafkaClient(config kube.Config, release helmette.Release, partial redpanda.PartialValues, dialer DialContextFunc, opts ...kgo.Opt) (*kgo.Client, error) {
+	dot, err := redpanda.Dot(release, partial)
```

Comment:
Should this function just accept a `Dot` instead of constructing it internally?

Reply:
Probably, it'd be a lot cleaner initialization to do so, I'll change it.